### PR TITLE
Implement structural equivalence for processes in RhoLang.

### PR DESCRIPTION
**Pull request by echatav**
This adds a module `Equivalence.scala` including name equivalence of channels and structural equivalence of processes and supporting functionality. The implementation is based on [rhocaml](https://github.com/leithaus/rhocaml/blob/master/rho.ml). Differences between basic Rho calculus and RhoLang necessitate some changes. For instance, `Input`s of joined processes are handled similarly to `Par` in terms of considering associativity and commutativity of ordering. Another difference that still must be handled is the existence of channel variables, channels that aren't quoted processes.

### CODE REVIEW

**Comment by nashef** [comment]
Please make pull request titles a complete sentence. Also, please include a paragraph or so of description. What's included, how did you implement it, notes on important decisions you made during implementation, etc.

File: rholang/src/main/scala/rholang/rosette/Equivalence.scala [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+object Equivalences{
+
+  def nameEquivalent(env1: DeBruijn, n1: Chan, env2: DeBruijn, n2: Chan): Boolean = {
+    (n1, n2) match {
+      case (q1: CQuote, q2: CQuote) =>
+        q1.proc_ match {
+          case pdrop1: PDrop => nameEquivalent(pdrop1.chan_, n2)
+          case _ => q2.proc_ match {
+            case pdrop2: PDrop => nameEquivalent(n1, pdrop2.chan_)
+            case _ => structurallyEquivalent(q1.proc_, q2.proc_)
```
Comment (guardbotmk3) [inline rholang/src/main/scala/rholang/rosette/Equivalence.scala]: This needs to propagate the environment.

File: rholang/src/main/scala/rholang/rosette/Equivalence.scala [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+import coop.rchain.syntax.rholang.Absyn._
+import scalaz.{Bind => _, Value => _, _}
+import scalaz.std.list._
+import scalaz.std.option._
+import scala.collection.JavaConverters._
+
+object Equivalences{
+
+  def nameEquivalent(env1: DeBruijn, n1: Chan, env2: DeBruijn, n2: Chan): Boolean = {
```
Comment (rabbitonweb) [inline rholang/src/main/scala/rholang/rosette/Equivalence.scala]: for readability:

```
  def nameEquivalent(env1: DeBruijn, n1: Chan, env2: DeBruijn, n2: Chan): Boolean = (n1, n2) match {
  case (q1: CQuote, q2: CQuote) =>
    (q1.proc_, q2.proc_) match {
      case (pdrop1: PDrop, _) => nameEquivalent(pdrop1.chan_, n2)
      case (_, pdrop2: PDrop) => nameEquivalent(n1, pdrop2.chan_)
      case _                  => structurallyEquivalent(env1,q1.proc_,env2,q2.proc_)
    }
  case (q1: CVar, q2: CVar) => env1.equivalent(q1.var_, env2, q2.var_)
  case _ => false
}
```

Comment (echatav) [reply rholang/src/main/scala/rholang/rosette/Equivalence.scala]: I don't really like lining up arrows usually because then when you make edits to cases, like say adding a new case, you have to edit other lines just to line things up, and you don't see in the diff only logic changes.

Comment (echatav) [reply rholang/src/main/scala/rholang/rosette/Equivalence.scala]: That said, if we want to just use `scalafmt` so there are no style fights, I'm down with that.

**Comment by dckc** [comment]
Thanks for the pointer to rhocaml! Here's hoping for time to try porting it to idris in [RhoCalulus.idr](https://github.com/dckc/togl/blob/master/RhoCalulus.idr).

**Comment by echatav** [comment]
@dckc `foldLeft` is used 3 times and `foldLeftM` once. I would guess that the use of `foldLeft` is data dependent in that it has to check for equivalence under different orderings, but I am curious how you might optimize that.

**Comment by dckc** [comment]
I was just thinking that if the fold is actually associative and commutative, it's just like summing up a bunch of numbers or any other reduce job... https://stackoverflow.com/questions/35675466/reductions-in-parallel-in-logarithmic-time

File: rholang/src/main/scala/rholang/rosette/Equivalence.scala (hunk 2 above)
Comment (rabbitonweb) [reply rholang/src/main/scala/rholang/rosette/Equivalence.scala]: My comment was not regarding the arrow alignment but dealing with pattern matching within pattern matching that was within pattern matching

File: rholang/src/main/scala/rholang/rosette/Equivalence.scala [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+            (rejects, bnd) => rejects match {
+              case (false, names1, names2, r, l) =>
+                allBindsEquivalent(env1, r ++ l ++ eqtl, env2, tail) match {
+                  case Some((name1,name2)) => (
+                    true,
+                    name1 ++ names1,
+                    name2 ++ names2,
+                    r, l
+                  )
+                  case None => l match {
```
Comment (rabbitonweb) [inline rholang/src/main/scala/rholang/rosette/Equivalence.scala]: again, consider flatting those pattern matching expressions, for example
```
case None if l.isEmpty =>
case None => 
```

File: rholang/src/main/scala/rholang/rosette/Equivalence.scala [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+      case Nil => b2 match {
+        case Nil => Some((Nil,Nil))
+        case _ => None
+      }
+      case head :: tail => b2.partition {
+        bind => !(bindEquivalent(env1,bind,env2,head).isEmpty)
+      } match {
+        case (Nil,_) => None
+        case (eqhd,eqtl) =>
+          val init = (
```
Comment (rabbitonweb) [inline rholang/src/main/scala/rholang/rosette/Equivalence.scala]: I think it is worth providing a structure here for readability & maintainability purposes. This tuple really tells my anything. I have to navigate whole fold to understand difference between `List[CPattern]` on `_2` and `_3`.

File: README.md [hunk 5]
```diff
@@ -4,6 +4,18 @@
...
+The `comm` subproject contains code for network related operations for RChain.
+
+The network layer is the lowest level component in the architecture and it
+is featured in our **Node.Hello (v0.1) release**. The simplest way to get
+started is with [docker][]: `docker run -ti rchain/rchain-comm`. For other options,
+see [comm/README.md][cr].
+
+[docker]: https://store.docker.com/community/images/rchain/rchain-comm
+[cr]: https://github.com/rchain/rchain/tree/master/comm
+
```
Comment (guardbotmk3) [inline README.md]: Unrelated change. Please don't do this.

File: rholang/src/main/scala/rholang/rosette/Equivalence.scala [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+            (cases1,cases2).zipped.forall {
+              case (pm1: PatternMatch,pm2: PatternMatch) =>
+                val pmp1 = pm1.proc_
+                val pmp2 = pm2.proc_
+                if (structurallyEquivalent(env1,pmp1,env2,pmp2)) {
+                  val pat1 = pm1.ppattern_
+                  val pat2 = pm2.ppattern_
+                  (pat1,pat2) match {
+                    case (val1: PPtVal, val2: PPtVal) =>
+                      val1.valpattern_ == val2.valpattern_
```
Comment (guardbotmk3) [inline rholang/src/main/scala/rholang/rosette/Equivalence.scala]: When you go to port this to the mercury grammar, you'll not that there are no patterns. This is on purpose, because a pattern *is* a process with (0 or more) free variables/wildcards.

File: rholang/src/main/scala/rholang/rosette/Equivalence.scala [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+              case tuple => tuple
+            }
+          } match {
+            case (false,_,_,_,_) => None
+            case (true,names1,names2,_,_) => Some((names1,names2))
+          }
+        }
+    }
+
+  def alphaEquivalent(p1: Proc, p2: Proc): Boolean = ???
```
Comment (guardbotmk3) [inline rholang/src/main/scala/rholang/rosette/Equivalence.scala]: You don't need this, that's what the environments are for. (If the vars represent the same Debruijn level, they are alpha equivalent)
Substitution is easier in rholang than rho calculus because we have patterns and don't use names as variables.

File: rholang/src/main/scala/rholang/rosette/Equivalence.scala [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+    case (p: PPar) => parLeaves(p.proc_1) ++ parLeaves(p.proc_2)
+  }
+
+  def parEquiv(procs1: List[Proc], procs2: List[Proc]): Boolean = procs1 match {
+    case Nil => procs2.forall(proc => structurallyNil(proc))
+    case head :: tail =>
+      procs2.partition(proc => structurallyEquivalent(head, proc)) match {
+        case (Nil, tl) => false
+        case (eqhd, eqtl) =>
+          eqhd.foldLeft((false, List[Proc](), eqhd.tail)) {
```
Comment (guardbotmk3) [inline rholang/src/main/scala/rholang/rosette/Equivalence.scala]: I'm not sure this is necessary. This looks like the algorithm for model checking rather than the algorithm for structural equivalence.

Because structural equivalence is an equivalence, you don't need to fold over the list of equivalent processes, any one of them is as good as any other.

File: rholang/src/main/scala/rholang/rosette/Equivalence.scala [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+                  }
+                }
+              case (true, r, l) => (true, r, l)
+            }
+          } match { case (eq,_,_) => eq }
+      }
+  }
+}
+
+class DeBruijn(val environment: Map[String,Int], val next: Int){
```
Comment (guardbotmk3) [inline rholang/src/main/scala/rholang/rosette/Equivalence.scala]: I used something similar in my normalizer. This was a good idea.

**Comment by echatav** [comment]
Closing as this will be subsumed by rholang-mercury
